control-plane/agents/src/bin/core/tests/rebuild/mod.rs

```diff
@@ -0,0 +1,260 @@
+        if r.uri == testrep.uri {
+            assert!(r.state == ReplicaStatus::Online.into());
+            return true;
+        }
+        false
+    });
+
+    // Get the volume again for validations.
+    let vol = volume_api.get_volume(volid.uuid()).await.unwrap();
+    let volume_state = vol.state;
+    let nexus = volume_state.target.unwrap();
+    // The child must be still in the nexus.
+    let is_removed = !nexus.children.iter().any(|c| c.uri == testrep.uri);
+    assert!(!is_removed);
+    // TODO: We can additionally check that the volume state eventually becomes Online,
+    // and doesn't remain Degraded once the partial rebuild finishes. Currently no support for
+    // getting rebuild info via nexus grpc client NexusOperations.
+}
+
+// This test
+// 1. creates a volume with 3 replicas, faults one of them by stopping container.
+// 2. Expects child's replica to stay offline for faulted child wait period.
+// 3. Since replica doesn't come online within wait period, child gets removed from nexus.
+// 4. Validate child should not be present in nexus.
+#[tokio::test]
+async fn replica_wait_and_full_rebuild() {
+    let cluster = build_cluster(4, 52428800, 0).await; // 50MiB pool size
+
+    let vol_target = cluster.node(0).to_string();
+    let api_client = cluster.rest_v00();
+    let volume_api = api_client.volumes_api();
+
+    let body = CreateVolumeBody::new(VolumePolicy::new(true), 3, 10485760u64, false);
+    let volid = VolumeId::new();
+    let volume = volume_api.put_volume(&volid, body).await.unwrap();
+    let volume = volume_api
+        .put_volume_target(
+            &volume.spec.uuid,
+            PublishVolumeBody::new_all(
+                HashMap::new(),
+                None,
+                vol_target.clone().to_string(),
+                models::VolumeShareProtocol::Nvmf,
+                None,
+                cluster.csi_node(0),
+            ),
+        )
+        .await
+        .unwrap();
+
+    let volume_state = volume.state;
+    assert!(volume_state.status == VolumeStatus::Online.into());
+    let nexus = volume_state.target.unwrap();
+    assert_eq!(nexus.children.len(), 3);
+    nexus
+        .children
+        .iter()
+        .for_each(|c| assert!(c.state == ChildState::Online.into()));
+
+    // Check where the replicas are, apart from vol target node.
+    let replicas = api_client.replicas_api().get_replicas().await.unwrap();
+    tracing::info!({?replicas}, "Available replicas - ");
+    let testrep = replicas.iter().find(|r| r.node != vol_target).unwrap();
+    let testrep_node = &testrep.node;
+    tracing::info!(
+        "Stopping node {testrep_node} having replica {}",
+        testrep.uri.clone()
+    );
+
+    // Stop the container and don't restart.
+    cluster
+        .composer()
+        .stop(&testrep_node.to_string())
+        .await
+        .expect("container stop failure");
+    // Give reconciler time to pitch-in.
+    sleep(Duration::from_secs(5));
```

Comment:
What are we waiting for here? Should probably have some loop checking for a condition.
Also should not use blocking sleeps on async functions.

Reply:
I can remove this and put a looping check on container state.